### PR TITLE
Added Grave Escape

### PR DESCRIPTION
**Pull request by jamesofarrell**
Adds GRAVE_ESCAPE and KC_GESC. This will act as escape when pressed but when pressed with shift will send a ~

code originally found here: https://github.com/p3lim/keyboard_firmware

### CODE REVIEW

**Comment by Wilba6582** [comment]
`GRAVE_ESC      = 0x5900,`  after `RESET = 0x5C00,` changes the int values of enums following it.

If it is to be added, it should go at the end of the list, just before `SAFE_RANGE`

IMHO it's a very specific use-case that probably doesn't belong in core QMK - it's something that could be entirely implemented in a user's keymap. The `quantum_keycodes` enum is already bloated and we should be wary of "feature creep" here.

**Comment by jamesofarrell** [comment]
Fair enogugh, feature creep is a problem I've jut noticed that grave/escape is on the default layout of a lot of 60% and 65% keyboards but it not easy to add in QMK without a macro.

**Comment by skullydazed** [comment]
I do see this in a lot of places, and having it standardized would be better than having it littered among a bunch of different keymaps. Keeping it up to date within my own keyboard is problematic, for example.

Before we merge this I think we should resolve a question around behavior here. I find that [checking for all mods](https://github.com/qmk/qmk_firmware/blob/master/keyboards/clueboard/keymaps/default/keymap.c#L4) is most useful to me, so I don't break Cmd-~ in OS X. I do see a lot of people using only shift, and I don't know if that's because they specifically want only shift, or if they simply haven't considered other mods as well. Can everyone live with all mods, or do we need to find a way to configure that behavior?

File: quantum/quantum.c [hunk 1]
```diff
@@ -437,6 +437,16 @@ bool process_record_quantum(keyrecord_t *record) {
       return false;
       // break;
     }
+    case GRAVE_ESC: {
+      void (*method)(uint8_t) = (record->event.pressed) ? &add_key : &del_key;
+      uint8_t shifted = get_mods() & (MOD_BIT(KC_LSHIFT) | MOD_BIT(KC_RSHIFT));
+
+      if(layer_state == 0)
```
Comment (tmk) [inline quantum/quantum.c]: why is this layer check needed? This seems to limit its usage to the layer condition exactly.

Comment (jamesofarrell) [inline quantum/quantum.c]: It is probably not needed but I think skullydazed linked some better code so i'll close this pull request, update my repo with the code from skullydazed and resubmit.

**Comment by jamesofarrell** [comment]
Looks like this code needs more work. I'll close the pull request when i get home in a few hours (incase anyone else has anymore suggestion). update the code so it uses the action from skullydazed. I'll also fix the place i put the new keycode. Thanks for the feedback everyone!

Once I've done that and tested it on a keyboard I'll resubmit the pull requests.

**Comment by jackhumbert** [comment]
Cool! We're also messing around with making these a little easier to implement for other uses - as you mentioned, folks have been wanting to do this for a while, so I think it's time we gave it some attention :) thanks for the jumping-off point!

**Comment by skullydazed** [comment]
I'll second what @jackhumbert  said. Also, no need to open a new PR, just push changes to your current branch when you're ready.

**Comment by Wilba6582** [comment]
I guess I was wrong about it being too specific... didn't know about Shift-Command-Tilde in OSX ;-)

**Comment by jamesofarrell** [comment]
I've made the suggestion changes but I've removed the detection for ALT as it would break control-alt-escape on windows.

**Comment by skullydazed** [comment]
Thanks!

**Comment by belak** [comment]
Just a random comment, but if the description is correct isn't "grave-escape" the wrong name here, since it'll send a tilde?

**Comment by Wilba6582** [comment]
It can't send a tilde... it can only send a grave keycode. A tilde is what the host might produce if the shift key is held at the same time.

**Comment by belak** [comment]
Ah, I see... my confusion was around the fact that this is mostly intended to be used with shift... anyway, thanks for the clarification.

**Pull request merged**